kernel/kernel-api/src/main/java/io/delta/kernel/internal/util/SchemaUtils.java

```diff
@@ -581,43 +638,84 @@ private static void validateFieldCompatibility(
+      // elements that have the same path but different types then the following scenarios are
+      // handled in this block.
+      // 1. This is non-leaf node (e.g. StructType, ArrayType) type, in which case it is sufficient
+      // to ensure that the types are of the same class. For a struct type there might be field
+      // additions or removals, which shouldn't be considered invalid schema transitions
+      // (and hence `equivalent(...)` is not used in this case). If the nested type changes
+      // (e.g. ArrayType to Primitive) then the classes would be different and the types by
+      // definition would not be equivalent.
+      //
+      // 2. This is a leaf node, in which case it sufficient to check that the types are equivalent
+      // (or once implemented the transition of types is valid).
+      //
+      // The subtle point here is for any non-leaf node change, the computed changes will include at
+      // least one ancestor change where the type change can be detected.
+      //
+      for (Class<?> clazz : new Class<?>[] {StructType.class, ArrayType.class}) {
+        if (existingField.getDataType().getClass() == clazz
+            && newField.getDataType().getClass() == clazz) {
+          return;
+        }
+      }
 
-      validateFieldCompatibility(
-          existingMapType.getKeyField(), newMapType.getKeyField(), icebergWriterCompatV1Enabled);
-      validateFieldCompatibility(
-          existingMapType.getValueField(),
-          newMapType.getValueField(),
-          icebergWriterCompatV1Enabled);
-    } else if (existingField.getDataType() instanceof ArrayType
-        && newField.getDataType() instanceof ArrayType) {
-      ArrayType existingArrayType = (ArrayType) existingField.getDataType();
-      ArrayType newArrayType = (ArrayType) newField.getDataType();
-
-      validateFieldCompatibility(
-          existingArrayType.getElementField(),
-          newArrayType.getElementField(),
-          icebergWriterCompatV1Enabled);
-    } else if (!existingField.getDataType().equivalent(newField.getDataType())) {
-      throw new KernelException(
-          String.format(
-              "Cannot change the type of existing field %s from %s to %s",
-              existingField.getName(), existingField.getDataType(), newField.getDataType()));
+      if (!existingField.getDataType().equivalent(newField.getDataType())) {
+        throw new KernelException(
+            String.format(
+                "Cannot change the type of existing field %s from %s to %s",
+                existingField.getName(), existingField.getDataType(), newField.getDataType()));
+      }
+    }
+  }
+
+  /**
+   * A composite class that uniquely identifiers an element in a schema.
+   *
+   * <p>When column mapping is enabled, the every field in structs has a unique ID. For other nested
+   * types (maps and arrays), elements are identified from there path from the struct field.
```

Comment:
With icebregCompatV2, map/array elements will have an ID defined as well. I think we can later extend this to be able to include that in the fieldID mapping? Not neccessary now though I think

Reply:
Yeah, I think it is question whether we want to special case this.  We also at some point need to handle things when column mappings is not enabled in general.